### ISSUE TITLE
Bug: @typescript-eslint/no-unused-vars does not work correct

### ISSUE THREAD

**Issue by 3axap4eHko** [typescript-eslint/typescript-eslint#10399]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgFWAGcYAefOUVJAEyLgFckBrJCAdyQG0BdAGjgAlAHxwA3gFgAUHDgAKAHRLMUAOZEAXAQCUWwQG5pAX2nTQkWHEwAjElBzxsAG0xF6hEuQEjKIanTgAMSZsGAQIJHE4EykgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export interface Test<T extends unknown[], R> {
  (...args: T): R;
}

export abstract class Test<T, R> extends Function {}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": 2,
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

First of all T and R are not variables, so it is unclear why they are reported. Also, the rule does not understand declaration merging, as you can see the interface Test uses T and R.

### Actual Result

error  'T' is defined but never used  @typescript-eslint/no-unused-vars
error  'R' is defined but never used  @typescript-eslint/no-unused-vars

**Comment by kirkwaiblinger** [comment]
Hi @3axap4eHko ! 

This is a known limitation of the rule, see https://github.com/typescript-eslint/typescript-eslint/issues/9811, and others.

You'll want to eslint-disable in this case.
